File: backend/api/v1/endpoints/groups.py
"""
Phase K2 — Learning groups endpoints.

Minimal viable surface: browse, create, join, leave, detail. Group
posts and scheduled-sessions live in their own tables (`group_posts`,
`group_sessions`) and are out of scope for this iteration — the data
model can be extended without breaking these endpoints.
"""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Body, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from core.security import get_current_active_user
from database.models import User
from database.supabase_client import SupabaseREST

logger = logging.getLogger(__name__)
router = APIRouter()


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class CreateGroupPayload(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    description: Optional[str] = None
    subject_id: Optional[str] = None
    group_type: str = "study"
    level: Optional[str] = None
    max_members: Optional[int] = None
    language: Optional[str] = "en"
    tags: Optional[List[str]] = None


@router.get("/students/groups")
async def list_groups(
    q: Optional[str] = Query(None),
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    """Browse all active learning groups."""
    rows = SupabaseREST.select(
        "learning_groups", "*", None, order="created_at.desc", limit=100
    ) or []
    if q:
        ql = q.lower()
        rows = [
            r for r in rows
            if ql in (r.get("name") or "").lower()
            or ql in (r.get("description") or "").lower()
        ]
    user_id = str(current_user.id)
    memberships = SupabaseREST.select(
        "group_memberships",
        "group_id,is_active",
        {"user_id": user_id, "is_active": True},
    ) or []
    member_ids = {str(m.get("group_id")) for m in memberships}
    for r in rows:
        r["is_member"] = str(r.get("id")) in member_ids
    return {"success": True, "groups": rows}


@router.get("/students/groups/mine")
async def list_my_groups(
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    """The groups the current user is in (admin or member)."""
    user_id = str(current_user.id)
    memberships = SupabaseREST.select(
        "group_memberships",
        "group_id,role,joined_at,is_active",
        {"user_id": user_id, "is_active": True},
    ) or []
    out: List[Dict[str, Any]] = []
    for m in memberships:
        g = SupabaseREST.select_one(
            "learning_groups", "*", {"id": str(m.get("group_id"))}
        )
        if g:
            g["my_role"] = m.get("role")
            g["joined_at"] = m.get("joined_at")
            out.append(g)
    return {"success": True, "groups": out}


@router.post("/students/groups")
async def create_group(
    payload: CreateGroupPayload,
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    """Create a new learning group. The creator is auto-enrolled as `admin`."""
    user_id = str(current_user.id)
    group_id = str(uuid.uuid4())
    now = _now_iso()

    inserted = SupabaseREST.insert(
        "learning_groups",
        {
            "id": group_id,
            "name": payload.name,
            "description": payload.description,
            "subject_id": payload.subject_id,
            "group_type": payload.group_type,
            "level": payload.level,
            "admin_id": user_id,
            "max_members": payload.max_members,
            "current_members": 1,
            "language": payload.language,
            "tags": payload.tags or [],
            "created_at": now,
        },
    )
    if not inserted:
        raise HTTPException(status_code=500, detail="Could not create group.")

    SupabaseREST.insert(
        "group_memberships",
        {
            "group_id": group_id,
            "user_id": user_id,
            "role": "admin",
            "joined_at": now,
            "is_active": True,
        },
    )

    return {"success": True, "group": inserted}


@router.get("/students/groups/{group_id}")
async def get_group(
    group_id: str,
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    g = SupabaseREST.select_one("learning_groups", "*", {"id": group_id})
    if not g:
        raise HTTPException(status_code=404, detail="Group not found.")

    members = SupabaseREST.select(
        "group_memberships",
        "user_id,role,joined_at",
        {"group_id": group_id, "is_active": True},
    ) or []
    member_rows: List[Dict[str, Any]] = []
    for m in members:
        p = SupabaseREST.select_one(
            "user_profiles",
            "first_name,last_name,avatar_url",
            {"user_id": str(m.get("user_id"))},
        ) or {}
        name = f"{p.get('first_name', '')} {p.get('last_name', '')}".strip() or "Member"
        member_rows.append(
            {
                "user_id": m.get("user_id"),
                "role": m.get("role"),
                "joined_at": m.get("joined_at"),
                "name": name,
                "avatar_url": p.get("avatar_url"),
            }
        )

    user_id = str(current_user.id)
    my_membership = next(
        (m for m in members if str(m.get("user_id")) == user_id),
        None,
    )

    return {
        "success": True,
        "group": g,
        "members": member_rows,
        "my_role": my_membership.get("role") if my_membership else None,
        "is_member": bool(my_membership),
    }


@router.post("/students/groups/{group_id}/join")
async def join_group(
    group_id: str,
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    user_id = str(current_user.id)
    group = SupabaseREST.select_one(
        "learning_groups", "id,max_members,current_members", {"id": group_id}
    )
    if not group:
        raise HTTPException(status_code=404, detail="Group not found.")

    existing = SupabaseREST.select_one(
        "group_memberships",
        "id,is_active",
        {"group_id": group_id, "user_id": user_id},
    )
    if existing:
        if existing.get("is_active"):
            return {"success": True, "already_member": True}
        SupabaseREST.update(
            "group_memberships",
            {"is_active": True, "joined_at": _now_iso()},
            {"id": str(existing["id"])},
        )
    else:
        if group.get("max_members") and (
            int(group.get("current_members") or 0) >= int(group["max_members"])
        ):
            raise HTTPException(status_code=400, detail="Group is full.")
        SupabaseREST.insert(
            "group_memberships",
            {
                "group_id": group_id,
                "user_id": user_id,
                "role": "member",
                "joined_at": _now_iso(),
                "is_active": True,
            },
        )

    SupabaseREST.update(
        "learning_groups",
        {"current_members": int(group.get("current_members") or 0) + 1},
        {"id": group_id},
    )
    return {"success": True}


@router.post("/students/groups/{group_id}/leave")
async def leave_group(
    group_id: str,
    current_user: User = Depends(get_current_active_user),
) -> Dict[str, Any]:
    user_id = str(current_user.id)
    existing = SupabaseREST.select_one(
        "group_memberships",
        "id,is_active,role",
        {"group_id": group_id, "user_id": user_id},
    )
    if not existing or not existing.get("is_active"):
        return {"success": True, "not_a_member": True}
    if existing.get("role") == "admin":
        raise HTTPException(
            status_code=400,
            detail="Admins cannot leave their own group. Transfer admin first.",
        )
    SupabaseREST.update(
        "group_memberships",
        {"is_active": False},
        {"id": str(existing["id"])},
    )
    group = SupabaseREST.select_one(
        "learning_groups", "current_members", {"id": group_id}
    ) or {}
    SupabaseREST.update(
        "learning_groups",
        {"current_members": max(0, int(group.get("current_members") or 1) - 1)},
        {"id": group_id},
    )
    return {"success": True}

File: backend_phoenix/lib/skillhub/groups.ex
defmodule SkillHub.Groups do
  @moduledoc "Learning groups context (groups.py): browse, mine, create, get, join, leave."
  import Ecto.Query, warn: false

  alias SkillHub.Repo
  alias SkillHub.Groups.{LearningGroup, GroupMembership}

  def list_groups(user_id, q) do
    groups =
      from(g in LearningGroup, order_by: [desc: g.created_at], limit: 100)
      |> Repo.all()
      |> filter_query(q)

    member_ids = active_group_ids(user_id)
    Enum.map(groups, fn g -> Map.put(to_map(g), :is_member, MapSet.member?(member_ids, g.id)) end)
  end

  def list_my_groups(user_id) do
    memberships =
      from(m in GroupMembership, where: m.user_id == ^user_id and m.is_active == true)
      |> Repo.all()

    by_group = Map.new(memberships, &{&1.group_id, &1})
    ids = Map.keys(by_group)

    from(g in LearningGroup, where: g.id in ^ids)
    |> Repo.all()
    |> Enum.map(fn g ->
      m = by_group[g.id]
      to_map(g) |> Map.put(:my_role, m.role) |> Map.put(:joined_at, m.joined_at)
    end)
  end

  def create_group(user_id, attrs) do
    group_id = Ecto.UUID.generate()

    params = %{
      id: group_id,
      name: attrs["name"],
      description: attrs["description"],
      subject_id: blank_to_nil(attrs["subject_id"]),
      group_type: attrs["group_type"] || "study",
      level: attrs["level"],
      admin_id: user_id,
      max_members: attrs["max_members"],
      current_members: 1,
      language: attrs["language"] || "en",
      tags: attrs["tags"] || []
    }

    with {:ok, group} <- %LearningGroup{} |> LearningGroup.create_changeset(params) |> Repo.insert() do
      %GroupMembership{}
      |> GroupMembership.changeset(%{group_id: group_id, user_id: user_id, role: "admin", is_active: true})
      |> Repo.insert()

      {:ok, group}
    end
  end

  def get_group(user_id, group_id) do
    case Repo.get(LearningGroup, group_id) do
      nil ->
        {:error, :not_found}

      group ->
        members =
          from(m in GroupMembership,
            where: m.group_id == ^group_id and m.is_active == true
          )
          |> Repo.all()

        member_rows =
          Enum.map(members, fn m ->
            p = profile(m.user_id)
            %{
              user_id: m.user_id,
              role: m.role,
              joined_at: m.joined_at,
              name: full_name(p) || "Member",
              avatar_url: p[:avatar_url]
            }
          end)

        mine = Enum.find(members, &(&1.user_id == user_id))

        {:ok,
         %{
           success: true,
           group: group,
           members: member_rows,
           my_role: mine && mine.role,
           is_member: not is_nil(mine)
         }}
    end
  end

  def join_group(user_id, group_id) do
    case Repo.get(LearningGroup, group_id) do
      nil ->
        {:error, :not_found}

      group ->
        existing =
          Repo.one(from m in GroupMembership, where: m.group_id == ^group_id and m.user_id == ^user_id)

        cond do
          existing && existing.is_active ->
            {:ok, %{success: true, already_member: true}}

          existing ->
            existing
            |> GroupMembership.changeset(%{is_active: true, joined_at: DateTime.utc_now()})
            |> Repo.update()

            bump_members(group_id, 1)
            {:ok, %{success: true}}

          group.max_members && (group.current_members || 0) >= group.max_members ->
            {:error, :full}

          true ->
            %GroupMembership{}
            |> GroupMembership.changeset(%{group_id: group_id, user_id: user_id, role: "member", is_active: true})
            |> Repo.insert()

            bump_members(group_id, 1)
            {:ok, %{success: true}}
        end
    end
  end

  def leave_group(user_id, group_id) do
    existing =
      Repo.one(from m in GroupMembership, where: m.group_id == ^group_id and m.user_id == ^user_id)

    cond do
      is_nil(existing) or not existing.is_active -> {:ok, %{success: true, not_a_member: true}}
      existing.role == "admin" -> {:error, :admin_cannot_leave}
      true ->
        existing |> GroupMembership.changeset(%{is_active: false}) |> Repo.update()
        bump_members(group_id, -1)
        {:ok, %{success: true}}
    end
  end

  # --- helpers ---------------------------------------------------------------

  defp to_map(%LearningGroup{} = g), do: g |> Map.from_struct() |> Map.drop([:__meta__])

  defp active_group_ids(user_id) do
    from(m in GroupMembership, where: m.user_id == ^user_id and m.is_active == true, select: m.group_id)
    |> Repo.all()
    |> MapSet.new()
  end

  defp bump_members(group_id, delta) do
    from(g in LearningGroup, where: g.id == ^group_id)
    |> Repo.update_all(inc: [current_members: delta])
  end

  defp filter_query(groups, nil), do: groups
  defp filter_query(groups, ""), do: groups
  defp filter_query(groups, q) do
    ql = String.downcase(q)
    Enum.filter(groups, fn g ->
      String.contains?(String.downcase(g.name || ""), ql) or
        String.contains?(String.downcase(g.description || ""), ql)
    end)
  end

  defp profile(user_id) do
    case Repo.query!(
           "select first_name, last_name, avatar_url from public.user_profiles where user_id = $1::uuid limit 1",
           [Ecto.UUID.dump!(user_id)]
         ).rows do
      [[first, last, avatar] | _] -> %{first_name: first, last_name: last, avatar_url: avatar}
      _ -> %{}
    end
  end

  defp full_name(%{first_name: f, last_name: l}) do
    name = String.trim("#{f || ""} #{l || ""}")
    if name == "", do: nil, else: name
  end
  defp full_name(_), do: nil

  defp blank_to_nil(nil), do: nil
  defp blank_to_nil(""), do: nil
  defp blank_to_nil(v), do: v
end
